llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -7503,6 +7503,17 @@ SDValue RISCVTargetLowering::lowerSELECT(SDValue Op, SelectionDAG &DAG) const {
       return DAG.getNode(ISD::ADD, DL, VT, CMOV, RHSVal);
     }
 
+    // c = setcc f, 0, seteq
+    // (select c, t, f) -> (or f, (czero_nez t, f))
```

Comment:
Proof: https://alive2.llvm.org/ce/z/7wUHBD

Reply:
Missing fold:
```
c = setcc t, 0, setne
(select c, t, f) -> (or t, (czero_nez f, t))
```
Proof: https://alive2.llvm.org/ce/z/3JcFbG

Reply:
Emm, it would be better to add the following folds in `RISCVInstrInfoZicond.td`:
```
def : Pat<(XLenVT (riscv_czero_eqz GPR:$rs1, (riscv_setne (XLenVT GPR:$rs1)))),
          (COPY GPR:$rs1)>;
def : Pat<(XLenVT (riscv_czero_nez GPR:$rs1, (riscv_seteq (XLenVT GPR:$rs1)))),
          (COPY GPR:$rs1)>;
```

Reply:
> Emm, it would be better to add the following folds in `RISCVInstrInfoZicond.td`:
> 
> ```
> def : Pat<(XLenVT (riscv_czero_eqz GPR:$rs1, (riscv_setne (XLenVT GPR:$rs1)))),
>           (COPY GPR:$rs1)>;
> def : Pat<(XLenVT (riscv_czero_nez GPR:$rs1, (riscv_seteq (XLenVT GPR:$rs1)))),
>           (COPY GPR:$rs1)>;
> ```

Sorry, it is incorrect.

Reply:
> Missing fold:
> 
> ```
> c = setcc t, 0, setne
> (select c, t, f) -> (or t, (czero_nez f, t))
> ```
> 
> Proof: https://alive2.llvm.org/ce/z/3JcFbG

Thanks. I have add the case.  BTW，do you still think it should be added in ` RISCVInstrInfoZicond.td` ?

Reply:
Yeah, it would be better to handle general cases :)
Can you try the following diff?

```
diff --git a/llvm/lib/Target/RISCV/RISCVISelLowering.cpp b/llvm/lib/Target/RISCV/RISCVISelLowering.cpp
index 539aa3525545..1960e7917080 100644
--- a/llvm/lib/Target/RISCV/RISCVISelLowering.cpp
+++ b/llvm/lib/Target/RISCV/RISCVISelLowering.cpp
@@ -16179,6 +16179,12 @@ SDValue RISCVTargetLowering::PerformDAGCombine(SDNode *N,
                            N->getOperand(0), Cond);
       }
     }
+    // czero_eqz x, (setcc x, 0, ne) -> x
+    // czero_nez x, (setcc x, 0, eq) -> x
+    if (N->getOperand(1).getOpcode() == ISD::SETCC && isNullConstant(N->getOperand(1).getOperand(1)) &&
+      cast<CondCodeSDNode>(N->getOperand(1).getOperand(2))->get() == (N->getOpcode() == RISCVISD::CZERO_EQZ ? ISD::CondCode::SETNE : ISD::CondCode::SETEQ)
+      && N->getOperand(0) == N->getOperand(1).getOperand(0))
+      return N->getOperand(0);
     return SDValue();
 
   case RISCVISD::SELECT_CC: {
```

Reply:
Thanks for your suggestion. I have changed the code in `RISCVTargetLowering: : PerformDAGCombine`.  It does get better.